File: mcp-ml-optimizer/mcp_ml_optimizer/classifier.py
# -*- coding: utf-8 -*-
"""
ML Classifier - Random Forest that maps error_code + message -> mcp_layer.

Reads training data from PostgreSQL, trains a model, and predicts
which MCP source file needs to be patched for a given error.
"""

import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.feature_extraction.text import TfidfVectorizer

from . import db

# Error code -> (category, mcp_layer) deterministic rules
RULES = {
    "TS2663": ("this_scope", "class-context-layer"),
    "TS2551": ("this_scope", "class-context-layer"),
    "TS2300": ("deduplication", "class-context-layer"),
    "NG8002": ("binding", "template-generator"),
    "NG8001": ("primeng_import", "primeng-mapper"),
    "NG5002": ("template_parse", "template-integrity-layer"),
    "NG9":    ("missing_property", "code-emitter"),
    "NG1":    ("missing_property", "code-emitter"),
    "NG8":    ("signal_type", "code-emitter"),
    "NG5":    ("type_safety", "code-emitter"),
    "TS2307": ("import_path", "code-emitter"),
    "TS2304": ("missing_name", "code-emitter"),
    "TS2306": ("import_path", "code-emitter"),
    "TS2571": ("signal_type", "code-emitter"),
    "TS18047":("type_safety", "code-emitter"),
    "TS2339": ("type_safety", "code-emitter"),
    "TS2345": ("type_safety", "code-emitter"),
    "TS2355": ("type_safety", "code-emitter"),
    "TS7006": ("type_safety", "code-emitter"),
    "TS7030": ("type_safety", "code-emitter"),
    "TS1005": ("syntax", "code-emitter"),
    "TS-991002": ("inline_template", "code-emitter"),
    "TS-992012": ("standalone_import", "code-emitter"),
}


class ErrorClassifier:
    def __init__(self):
        self.clf = None
        self.tfidf = None
        self.label_map = {}

    def train(self):
        """Train from PostgreSQL data. Returns True if model was trained."""
        rows = db.get_all_errors()
        if len(rows) < 5:
            print(f"  [ML] Only {len(rows)} samples - using rules only")
            return False

        texts = [f"{r['code']} {r['message'][:200]} {r['category']}" for r in rows]
        labels = [r['mcp_layer'] for r in rows]

        unique = sorted(set(labels))
        to_int = {l: i for i, l in enumerate(unique)}
        self.label_map = {i: l for l, i in to_int.items()}
        y = np.array([to_int[l] for l in labels])

        self.tfidf = TfidfVectorizer(max_features=200, ngram_range=(1, 2))
        X = self.tfidf.fit_transform(texts)

        self.clf = RandomForestClassifier(
            n_estimators=80, max_depth=8,
            random_state=42, class_weight="balanced",
        )
        self.clf.fit(X, y)
        acc = self.clf.score(X, y)
        print(f"  [ML] Trained on {len(rows)} samples, accuracy: {acc:.0%}")
        return True

    def classify(self, code, message=""):
        """Classify an error -> (category, mcp_layer, confidence)."""
        # Rule-based first
        if code in RULES:
            cat, layer = RULES[code]
            return cat, layer, 1.0

        # ML fallback
        if self.clf and self.tfidf:
            text = f"{code} {message[:200]}"
            proba = self.clf.predict_proba(self.tfidf.transform([text]))[0]
            idx = int(np.argmax(proba))
            return "unknown", self.label_map.get(idx, "unknown"), float(proba[idx])

        return "unknown", "unknown", 0.0

File: src/ml/classifier.ts
/**
 * Error Classifier — Rule-based + frequency scoring.
 *
 * Classifies TypeScript/Angular compilation errors into categories
 * and identifies which MCP layer needs to be patched.
 *
 * Strategy:
 *   1. Deterministic rules (hardcoded error codes) → confidence 1.0
 *   2. Prefix matching (e.g. NG8xxx) → confidence 0.9
 *   3. Frequency-based from DB history → confidence 0.7
 *   4. Unknown fallback → confidence 0.0
 */

import { getAllErrors } from './db-client.js';

// ─── Types ───────────────────────────────────────────────────────────────────

export interface ClassifyResult {
  category: string;
  mcpLayer: string;
  confidence: number;
}

// ─── Deterministic Rules ─────────────────────────────────────────────────────

/**
 * Maps error codes to [category, mcpLayer].
 * These cover 95%+ of cases deterministically.
 */
const RULES: Record<string, [string, string]> = {
  // Scope / this
  'TS2663': ['this_scope', 'class-context-layer'],
  'TS2551': ['this_scope', 'class-context-layer'],
  'TS2300': ['deduplication', 'class-context-layer'],

  // Template
  'NG8002': ['binding', 'template-generator'],
  'NG8001': ['primeng_import', 'primeng-mapper'],
  'NG5002': ['template_parse', 'template-integrity-layer'],

  // Component properties
  'NG9': ['missing_property', 'code-emitter'],
  'NG1': ['missing_property', 'code-emitter'],

  // Signals / types
  'NG8': ['signal_type', 'code-emitter'],
  'NG5': ['type_safety', 'code-emitter'],

  // Imports
  'TS2307': ['import_path', 'code-emitter'],
  'TS2304': ['missing_name', 'code-emitter'],
  'TS2306': ['import_path', 'code-emitter'],

  // Type safety
  'TS2571': ['signal_type', 'code-emitter'],
  'TS18047': ['type_safety', 'code-emitter'],
  'TS2339': ['type_safety', 'code-emitter'],
  'TS2345': ['type_safety', 'code-emitter'],
  'TS2355': ['type_safety', 'code-emitter'],
  'TS7006': ['type_safety', 'code-emitter'],
  'TS7030': ['type_safety', 'code-emitter'],

  // Syntax
  'TS1005': ['syntax', 'code-emitter'],

  // Custom codes
  'TS-991002': ['inline_template', 'code-emitter'],
  'TS-992012': ['standalone_import', 'code-emitter'],

  // PostCSS / Tailwind / Build config errors
  'POSTCSS_CONFIG': ['postcss_config', 'project-scaffolder'],
  'MODULE_NOT_FOUND': ['missing_dependency', 'project-scaffolder'],
  'SCSS_ERROR': ['style_error', 'project-scaffolder'],
};

// ─── Frequency Map (trained from DB) ────────────────────────────────────────

interface FreqEntry {
  category: string;
  mcpLayer: string;
  count: number;
}

let _freqMap: Map<string, FreqEntry> | null = null;

// ─── Public API ──────────────────────────────────────────────────────────────

/**
 * Train the frequency model from historical error data in the database.
 * Returns true if training succeeded (enough data), false otherwise.
 */
export async function train(): Promise<boolean> {
  try {
    const rows = await getAllErrors();
    if (rows.length < 5) return false;

    // Build frequency map: for each error code, find the most common category+layer
    const freq = new Map<string, Map<string, number>>();
    for (const r of rows) {
      if (!freq.has(r.code)) freq.set(r.code, new Map());
      const key = `${r.category}|${r.mcp_layer}`;
      const m = freq.get(r.code)!;
      m.set(key, (m.get(key) ?? 0) + 1);
    }

    _freqMap = new Map();
    for (const [code, layerCounts] of freq) {
      let best = '';
      let bestCount = 0;
      for (const [k, count] of layerCounts) {
        if (count > bestCount) { best = k; bestCount = count; }
      }
      const [category, mcpLayer] = best.split('|');
      _freqMap.set(code, { category, mcpLayer, count: bestCount });
    }

    return true;
  } catch {
    return false;
  }
}

/**
 * Classify a compilation error into a category and target MCP layer.
 *
 * @param code - Error code (e.g. "TS2663", "NG8001")
 * @param _message - Error message (reserved for future NLP-based classification)
 */
export function classify(code: string, _message = ''): ClassifyResult {
  // 1. Exact match in deterministic rules
  if (RULES[code]) {
    const [category, mcpLayer] = RULES[code];
    return { category, mcpLayer, confidence: 1.0 };
  }

  // 2. Prefix match (e.g. NG8xxx matches NG8)
  for (const ruleCode of Object.keys(RULES)) {
    if (code.startsWith(ruleCode)) {
      const [category, mcpLayer] = RULES[ruleCode];
      return { category, mcpLayer, confidence: 0.9 };
    }
  }

  // 3. Frequency-based from training data
  if (_freqMap?.has(code)) {
    const entry = _freqMap.get(code)!;
    return { category: entry.category, mcpLayer: entry.mcpLayer, confidence: 0.7 };
  }

  // 4. Unknown
  return { category: 'unknown', mcpLayer: 'unknown', confidence: 0.0 };
}
